rocketpy/simulation/monte_carlo.py

```diff
@@ -288,43 +279,54 @@ def __run_in_serial(self):
         -------
         None
         """
-        input_file = open(self._input_file, 'a', encoding="utf-8")
-        output_file = open(self._output_file, 'a', encoding="utf-8")
-        error_file = open(self._error_file, 'a', encoding="utf-8")
-
+        sim_monitor = _SimMonitor(
+            initial_count=self._initial_sim_idx,
+            n_simulations=self.number_of_simulations,
+            start_time=time(),
+        )
         try:
-            while self.__iteration_count < self.number_of_simulations:
-                self.__run_single_simulation(
-                    self.__iteration_count + self.num_of_loaded_sims,
-                    input_file,
-                    output_file,
+            while sim_monitor.keep_simulating():
+                sim_monitor.increment()
+
+                inputs_dict, outputs_dict = MonteCarlo.__run_single_simulation(
+                    sim_monitor.count,
+                    self.environment,
+                    self.rocket,
+                    self.flight,
+                    self.export_list,
+                    self.export_sample_time,
+                )
+
+                MonteCarlo.__export_flight_data(
+                    inputs_dict,
+                    outputs_dict,
+                    self._input_file,
+                    self._output_file,
                 )
-            print(
-                f"Completed {self.__iteration_count} iterations. Total CPU time: "
-                f"{process_time() - self.__start_cpu_time:.1f} s. Total wall time: "
-                f"{time() - self.__start_time:.1f} s\n"
-            )
+
+                sim_monitor.update_status(sim_monitor.count)
+
+            sim_monitor.finalize_status()
 
         except KeyboardInterrupt:
-            print("Keyboard Interrupt, files saved.")
-            error_file.write(json.dumps(self._inputs_dict, cls=RocketPyEncoder) + "\n")
+            MonteCarlo._reprint("Keyboard Interrupt, files saved.")
+            with open(self._error_file, "a", encoding="utf-8") as file:
```

Comment:
Any reason for enforcing utf-8 ? I was once told to not do that specially in an open-source library where people of different languages and operatoional systems could benefit form it.

Reply:
My main concern was standardization, since we might have to read these files later and knowing the encoding will make things easier. The file saving overall will be refactored with better encoding support, but there are generally two approaches to save files:

- Prioritize human readability / editing: in this case the encoding is important and is a good practice to have at least a standard. There are a plethora of writing systems in the world, of course `utf-8` does not cover everything, but it has become an "Internet standard". From [python docs](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files): "Because UTF-8 is the modern de-facto standard, encoding="utf-8" is recommended unless you know that you need to use a different encoding.".
- Prioritize I/O operations: this generally requires saving things as bytes, which does not require encoding. This makes all files "compatible" (reading simulations). However, they are generally not human readable and it is very hard to edit them in common editors.

I am more inclined to the current first option, which makes easier to read and make changes to the files.

Reply:
That's ok. I'm ok with this.